add DensityMatrix.to_statevector() for pure states

ajavadia opened the pull request:
This is handy when you have a circuit in a product state and you want the reduced state as a statevector (e.g. discarding an uncomputed ancilla).

jaygambetta commented on qiskit/quantum_info/states/densitymatrix.py:
```diff
@@ -585,6 +585,24 @@ def _evolve_instruction(self, obj, qargs=None):
+
+        eigvals, eigvecs = np.linalg.eig(self.data)
+        psi = eigvecs[:, np.argmax(eigvals)]  # eigenvector are in columns.
+        return Statevector(psi)
```
there should only be one  eigenvalue  and  you should check if it is one.  I thought there was a command to get the eigen decomposition (well i  used   it in qsphere) which would be better to use  and only accept if it has  one eigenvector and  equal  to one

jaygambetta commented on test/python/quantum_info/states/test_densitymatrix.py:
```diff
@@ -359,6 +359,22 @@ def test_to_dict(self):
+        psi = Statevector(state)
+        rho = DensityMatrix(psi)
+        phi = rho.to_statevector()
+        self.assertAlmostEqual(psi, phi)
```
while this  test will pass it  is not correct  due to  global phase i would would use the state fidelity

chriseclectic commented on qiskit/quantum_info/states/densitymatrix.py:
```diff
@@ -585,6 +585,33 @@ def _evolve_instruction(self, obj, qargs=None):
+        evals, evecs = np.linalg.eig(self._data)
+
+        nonzero_evals = evals[abs(evals) > atol]
+        if len(nonzero_evals) != 1 or not np.isclose(abs(nonzero_evals[0]), 1,
```
I think this should be
```suggestion
        if len(nonzero_evals) != 1 or not np.isclose(np.max(nonzero_evals), 1,
```

chriseclectic replied:
If you want to allow rank-1 projector density matrices that are not-trace 1 this could also be changed to
```suggestion
        if len(nonzero_evals) != 1 or not nonzero_evals[0] > 0:
```

chriseclectic commented on qiskit/quantum_info/states/densitymatrix.py:
```diff
@@ -585,6 +585,33 @@ def _evolve_instruction(self, obj, qargs=None):
+                                                     atol=atol, rtol=rtol):
+            raise QiskitError("Density matrix is not a pure state")
+
+        return Statevector(np.sqrt(np.max(evals)) * evals[:, np.argmax(evals)])
```
If you are restricting to the case where the eigenvalue is 1 as above you don't need to worry about rescaling here so can just have
```suggestion
        return Statevector(evals[:, np.argmax(evals)])
```

ajavadia commented:
@chriseclectic @jaygambetta addressed your comments.
Chris I restricted it to trace-1 density matrices for now as that is the `is_valid` definition currently.

chriseclectic commented on qiskit/quantum_info/states/densitymatrix.py:
```diff
@@ -585,6 +585,36 @@ def _evolve_instruction(self, obj, qargs=None):
+        Returns:
+            Statevector: The pure density matrix's corresponding statevector.
+                Corresponds to the eigenvector of the only non-zero eigenvalue.
+            atol (float): Absolute tolerance for checking operation validity.
```
I think these are supposed to be under `Args:` not `Returns:`
